mlflow/utils/requirements_utils.py

```diff
@@ -100,3 +112,109 @@ def _parse_requirements(requirements_file, is_constraint):
+
+
+_MODULE_TO_PACKAGES = importlib_metadata.packages_distributions()
+
+
+def _module_to_packages(module_name):
+    """
+    Returns a list of packages that provide the specified module.
+    """
+    return _MODULE_TO_PACKAGES.get(module_name, [])
+
+
+def _get_requires_recursive(pkg_name):
+    """
+    Recursively yields both direct and transitive dependencies of the specified package.
+    """
+    if pkg_name not in pkg_resources.working_set.by_key:
+        return
+
+    package = pkg_resources.working_set.by_key[pkg_name]
+    reqs = package.requires()
+    if len(reqs) == 0:
+        return
+
+    for req in reqs:
+        yield req.name
+        yield from _get_requires_recursive(req.name)
+
+
+def _prune_packages(packages):
+    """
+    Prunes packages required by other packages. For example, `["scikit-learn", "numpy"]` is pruned
+    to `["scikit-learn"]`.
+    """
+    requires = _flatten(map(_get_requires_recursive, packages))
+    return set(packages) - set(requires)
+
+
+def _run_command(cmd):
+    """
+    Runs the specified command.
+    """
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = proc.communicate()
+    stdout = stdout.decode("utf-8")
+    stderr = stderr.decode("utf-8")
+    if proc.returncode != 0:
+        msg = "\n".join(
+            [
+                f"Encountered an unexpected error while running {cmd}",
+                f"exit status: {proc.returncode}",
+                f"stdout: {stdout}",
+                f"stderr: {stderr}",
+            ]
+        )
+        raise MlflowException(msg)
+
+
+def _get_package_version(package):
+    """
+    Returns the version of the specified packages.
+    """
+    version = importlib_metadata.version(package)
+    # Strips a dev version suffix (e.g. '.dev0') for pyspark in Databricks.
+    if package == "pyspark" and is_in_databricks_runtime():
+        version = _strip_dev_version_suffix(version)
+    return version
+
+
+def _infer_requirements(model_uri, flavor):
+    """
+    Infers the pip requirements of the specified model.
+    """
+    local_model_path = _download_artifact_from_uri(model_uri)
+
+    # Run `_capture_modules.py` to capture modules imported during the loading procedure
+    with tempfile.TemporaryDirectory() as tmpdir:
+        output_file = os.path.join(tmpdir, "output.txt")
+        _run_command(
+            [
+                sys.executable,
+                _capture_modules.__file__,
+                "--model-path",
+                local_model_path,
+                "--flavor",
+                flavor,
+                "--output-file",
+                output_file,
+            ]
+        )
+        with open(output_file) as f:
+            modules = f.read().splitlines()
+
+    packages = _flatten(map(_module_to_packages, modules))
+    packages = map(_canonicalize_package_name, packages)
+    packages = set(packages) - set(["setuptools", "pkg_resources"])
```

Comment:
Can we add a comment explaining why we remove these?

Reply:
Sure!